Use libc::getrandom on Solaris and update docs.

josephlr opened the pull request:
https://github.com/rust-random/getrandom/pull/417 used `getentropy(2)` on Solaris, but after looking at [the blog post introducing `getrandom()` and `getentropy()`](https://blogs.oracle.com/solaris/post/solaris-new-system-calls-getentropy2-and-getrandom2), it seems like we should prefer using `getrandom` based on this quote:

> On Solaris the output of getentropy(2) is entropy and should not be used where randomness is needed, in particular it must not be used where an IV or nonce is needed when calling a cryptographic operation.  It is intended only for seeding a user space RBG (Random Bit Generator) system. More specifically the data returned by getentropy(2) has not had the required FIPS 140-2 processing for the DRBG applied to it.

I also updated some of the documentation explaining:
  - Why we ever use `getentropy(2)`
  - Why we set `GRND_RANDOM` on Solaris
  - Why we don't ever set `GRND_RANDOM` on other platforms.

newpavlov commented:
Left comment [here](https://github.com/rust-random/getrandom/pull/417#issuecomment-2095341300) just before this PR was created.

josephlr commented:
> Left comment [here](https://github.com/rust-random/getrandom/pull/417#issuecomment-2095341300) just before this PR was created.

Sounds good! I'll put Solaris in its own file.

josephlr commented:
@newpavlov I moved Solaris stuff to its own file. I also changed the implementation to always set `GRND_RANDOM`. 

Looking over the man pages, I think your comment in https://github.com/rust-random/getrandom/pull/417#issuecomment-2095341300 (noting that all "correct" examples use `GRND_RANDOM`) is a good point. I don't think it hugely matters one way or another, and if we are doing a special implementation for Solaris, we might as well just mimic what they recommend in their man pages.

Let me know if you want me to change it back to using `flags = 0` for consistency with the other uses of `getrandom(2)`.

newpavlov commented on src/solaris.rs:
```diff
@@ -0,0 +1,28 @@
+        let ret = unsafe { libc::getrandom(ptr, chunk.len(), libc::GRND_RANDOM) };
+        if ret != (chunk.len() as isize) {
+            return Err(last_os_error());
+        }
```
It also may be worth to add a note that `getrandom_inner` never receives an empty buffer.

josephlr replied:
Fixed, but I made the error check `ret < 0`, as it's negative values which indicates errors for `getrandom`.

josephlr replied:
Oh weird it seems like the docs contradict themselves:

"Upon successful completion, the getrandom() function returns the number of bytes written to buf. Otherwise, it returns 0 and sets errno to indicate the error."

"the function returns -1 and errno is set to EAGAIN."

newpavlov replied:
No, the check should be `ret == 0` as per the cited man quote (yeah, I was surprised as well...).

josephlr replied:
I think checking for both negative and zero return values is probably the safe thing to do here. I'm shocked how weird the solaris implementation of getrandom is, considering how normal the Illumos one is.

newpavlov replied:
>the function returns -1 and errno is set to EAGAIN.

IIUC it only applies to `GRND_NONBLOCK`.

josephlr replied:
Updated to check for negative and zero values.

newpavlov replied:
It could be worth to slightly modify the check to have only one branch in the happy path:
```rust
if ret as usize != chunk.len() {
    return if ret == 0 || ret == -1 {
        last_os_error()
    } else {
        Err(Error::UNEXPECTED)
    };
}
```

josephlr replied:
I think that is harder to read (and will probably get compiled to the same thing regardless).

newpavlov replied:
You can tweak formatting to your liking.

>will probably get compiled to the same thing regardless

It does not: https://rust.godbolt.org/z/G6d1zq4hs But it's a small enough difference, so I do not insist on this change.

josephlr replied:
Oh interesting. I'll leave it as-is, but that's good to know.
